core/server/master/src/test/java/alluxio/master/block/BlockMasterTest.java

```diff
@@ -380,10 +416,130 @@ public void getBlockInfo() throws Exception {
     assertEquals(expectedBlockInfo, mBlockMaster.getBlockInfo(blockId));
   }
 
+  /**
+   * RW contention
+   * Concurrent commit and readers
+   * Signal in commit and the readers inquire the state
+   * */
+  @Test
+  public void concurrentCommitWithReaders() throws Exception {
```

Comment:
@apc999 This is the test I wanted to discuss about. In this PR I change the locking in `DefaultBlockMaster` from a big `synchronized` section to using RW locks. So the testing on RW and WW contentions are more necessary than before.

In short the goal is to deterministically trigger an R or W2 operation while the W1 operation is running. In this way the reader or writer will test if W1 operation is protecting the internal metadata properly.

The `SignalBlockMaster` extends the `DefaultBlockMaster` but sends a signal in the middle of a write operation. The signal is sent on `createJournalContext()` because both `commit` and `remove` will call this in the middle, and this is part of the public API for the `BlockMaster` so it's better than hacking a private method.

Reply:
Do you think this kind of injection is acceptable? The interface is mostly intact. The cons might be: (1) I removed `final` on `DefaultBlockMaster`; (2) `workerRegister` and `workerHeartbeat` in `DefaultBlockMaster` cannot be injected the same way because they don't create a journal context.

If not, are there better ways?

Reply:
- removing `final` is actually good. I would even suggest removing most `final` modifiers in our code base for important classes.
- another way you can do is to overriding `alluxio.master.block.DefaultBlockMaster#lockBlock` which returns a modified `LockResource` --- in your modified `LockResource`,  you can signal other threads and wait something to happen before calling `unlock`

Reply:
@apc999 My only concern about overriding `alluxio.master.block.DefaultBlockMaster#lockBlock` is, that's a private method. I can make it inheritable, of course, but is inheriting a public `createJournalContext` more elegant than `lockBlock` or they are actually the same? Is it fine to rely on the fact that `remove` is calling either `lockBlock` or `createJournalContext`? Would you consider the test rely on method internals?

Other than that, inheriting `lockBlock` is a better choice because that enables me to throw signals from worker register and heartbeat.

Reply:
Changed to throw a signal out of `lockBlock`